### PR TITLE
Fix token approvals and clarify fields "protocolId", "locator", and "subject"

### LINKED ISSUE #661: Synchronous token approval timeout
Currently, synchronous token approvals will time out when submitting an approval for a pool that already has an existing token approval. This is because the `localID` assigned to the sync request has the possibility to be overwritten with the ID of the existing token approval because token approvals are mutable.

Scenario:
1. A token pool `123` exists with an existing approval `X`
2. A new approval `Y` is submitted for pool `123` and assigned a local ID.
3. When FireFly receives the approval event for `Y`, it determines that there is an existing approval `X` with the same `connector`, `pool`, and `protocolID` attributes.
4. FireFly updates approval `X` with the attributes from `Y`.
5. Approval `Y`'s local ID is discarded, so the FireFly sync/async bridge is stuck waiting.

### PR DESCRIPTION
**Pull request by awrichar**
Depends on https://github.com/hyperledger/firefly-tokens-erc1155/pull/67 and https://github.com/hyperledger/firefly-tokens-erc20-erc721/pull/42

This started as a quest to clean up token approvals, but ended up encompassing a few related items:

* The term "protocolId" has come to mean (fairly specifically) "an ID meaningful in the context of an underlying blockchain protocol".
  * The existing "protocolId" of token approvals is renamed to "subject", and token approvals get a new "protocolId" field that better aligns with the above.
  * Token transfers keep their "protocolId" unchanged, as it already aligns.
  * The existing "protocolId" of token pools is renamed to "locator".
  * Blockchain events will be de-duplicated based on their "protocolId" - only one event per namespace+listener+protocolId will be recorded, to avoid recording the same blockchain event multiple times.
* Token approvals will record _all_ historical approvals (instead of sometimes overwriting old ones). As a convenience for query purposes, the most recent approval for each subject will be notated with a new field "active=true".
* When matching token approval (and transfer) events to operations, event manager will now consider the operation ID _and_ the token connector and pool embedded in the operation inputs before considering it a match. This ensures that side-effects in other pools will not be matched with the wrong operation and return unexpected results for synchronous actions (when using confirm=true). Resolves #661.

This does introduce changes in the interface with token connectors, so will require connectors to be upgraded alongside it.

### CODE REVIEW

File: internal/events/tokens_approved.go [hunk 1]
```diff
@@ -54,61 +56,71 @@ func (em *eventManager) loadApprovalOperation(ctx context.Context, tx *fftypes.U
...
-			}
-
-			if existingLocal != nil {
-				approval.LocalID = fftypes.NewUUID()
-			}
-		} else {
-			log.L(ctx).Infof("Updating existing approval with local ID: %s", existing.LocalID)
-			approval.LocalID = existing.LocalID
+		if existingLocal != nil {
+			approval.LocalID = fftypes.NewUUID()
```
Comment (peterbroadhurst) [inline internal/events/tokens_approved.go:93]: A comment here would be worthwhile. My understanding is:

```
// The approval was initiated by FireFly and one UUID was allocated as a `localId` for the approval when it was
// initiated. This allows the `confirm=true` code to know when the first approval record gets inserted into the
// database that was correlated to that approval. However, approvals can result in many approval records
// depending on the on-chain mapping of the scope of an approval. So the connector might generate multiple
// notifications (with different `protocolId`), but all will reach this code with the same `localId`.
// So we need to generate a new one (for all apart from the first)
```

**Review by peterbroadhurst — APPROVED** [review]
Looks great. Think this is a significant step forwards in consistency.

I also reviewed:
- https://github.com/hyperledger/firefly-tokens-erc1155/pull/67
- https://github.com/hyperledger/firefly-tokens-erc20-erc721/pull/42

I think those need merging, releases cutting, and a manifest update here before this merges.

There's one place I had to stare at the code for a while to work it out, so I've proposed a comment based on my understanding.

Marking approval, and leaving you to work through the mechanics of the merge.

File: internal/events/tokens_approved.go (hunk 1 above)
Comment (awrichar) [reply internal/events/tokens_approved.go:93]: Yep, I'll add some notes

Comment (awrichar) [reply internal/events/tokens_approved.go]: Added a lot of notes and also cleaned up the method further to (hopefully) make it clearer to follow.

**Pull request merged**